### PR TITLE
Stabilize expr_2021 fragment specifier in all editions

### PR DESCRIPTION
**Pull request by eholk**
This is part of the `expr`/`expr_2021` fragment specifier for Edition 2024 (#123742). The RFC says we can support expr_2021 in as many editions as is practical, and there's nothing particularly hard about supporting it all the way back to 2015.

In editions 2021 and earlier, `expr` and `expr_2021` are synonyms. Their behavior diverges starting in Edition 2024. This is checked by the `expr_2021_inline_const.rs` test.

cc @vincenzopalazzo @rust-lang/wg-macros @traviscross

### CODE REVIEW

**Comment by traviscross** [comment]
@rustbot labels +I-lang-nominated

Thanks for the stabilization PR.  Let's nominate to have a look.

File: compiler/rustc_expand/src/mbe/quoted.rs [hunk 1]
```diff
@@ -13,10 +12,7 @@ use crate::errors;
 use crate::mbe::macro_parser::count_metavar_decls;
 use crate::mbe::{Delimited, KleeneOp, KleeneToken, MetaVarExpr, SequenceRepetition, TokenTree};
 
-const VALID_FRAGMENT_NAMES_MSG: &str = "valid fragment specifiers are \
-    `ident`, `block`, `stmt`, `expr`, `pat`, `ty`, `lifetime`, `literal`, `path`, `meta`, `tt`, \
-    `item` and `vis`";
-pub(crate) const VALID_FRAGMENT_NAMES_MSG_2021: &str = "valid fragment specifiers are \
+pub(crate) const VALID_FRAGMENT_NAMES_MSG: &str = "valid fragment specifiers are \
     `ident`, `block`, `stmt`, `expr`, `expr_2021`, `pat`, `ty`, `lifetime`, `literal`, `path`, \
     `meta`, `tt`, `item` and `vis`";
```
Comment (compiler-errors) [inline compiler/rustc_expand/src/mbe/quoted.rs:17]: It's weird to mention `expr_2021` and not for example mention `pat_param`. Can you reword this to:

```suggestion
pub(crate) const VALID_FRAGMENT_NAMES_MSG: &str = "valid fragment specifiers are \
    `ident`, `block`, `stmt`, `expr`, `pat`, `ty`, `lifetime`, `literal`, `path`, \
    `meta`, `tt`, `item` and `vis`, along with `expr_2021` and `pat_param` for edition compatibility";
```

**Comment by compiler-errors** [comment]
r=me after T-lang deals with this

File: compiler/rustc_expand/src/mbe/quoted.rs (hunk 1 above)
Comment (eholk) [reply compiler/rustc_expand/src/mbe/quoted.rs]: I like that suggestion, thanks!

**Comment by traviscross** [comment]
On the lang side, we last [discussed](https://github.com/rust-lang/rust/issues/123742#issuecomment-2315764431) this specifier on 2024-08-28 (in deciding to keep the name) and also noted in that discussion wanting to get this stabilized as soon as possible.  This is that stabilization, so let's...

@rfcbot fcp merge

**Comment by traviscross** [comment]
@rfcbot fcp merge

**Comment by tmandry** [comment]
@rfcbot reviewed

**Comment by vincenzopalazzo** [comment]
Small rebased and added the following diff!

```
diff --git a/tests/ui/macros/metavar_cross_edition_recursive_macros.rs b/tests/ui/macros/metavar_cross_edition_recursive_macros.rs
index 3eec1208b89..9a5b92f5032 100644
--- a/tests/ui/macros/metavar_cross_edition_recursive_macros.rs
+++ b/tests/ui/macros/metavar_cross_edition_recursive_macros.rs
@@ -6,7 +6,6 @@
 // This test captures the behavior of macro-generating-macros with fragment
 // specifiers across edition boundaries.

-#![feature(expr_fragment_specifier_2024)]
 #![feature(macro_metavar_expr)]
 #![allow(incomplete_features)]
```

Giving a CI run before squash the fixup commit!
